src/Microsoft.AspNet.Server.Kestrel/Http/Frame.Templated.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Hosting.Server;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public class Frame<THttpContext> : Frame
+    {
+        private readonly IHttpApplication<THttpContext> _application;
+
+        public Frame(IHttpApplication<THttpContext> application,
+                     ConnectionContext context)
+            : this(application, context, remoteEndPoint: null, localEndPoint: null)
+        {
+        }
+
+        public Frame(IHttpApplication<THttpContext> application,
+                     ConnectionContext context,
+                     IPEndPoint remoteEndPoint,
+                     IPEndPoint localEndPoint)
+            : base(context, remoteEndPoint, localEndPoint)
+        {
+            _application = application;
+        }
+
+        public override async Task RequestProcessingAsync()
+        {
+            try
+            {
+                var terminated = false;
+                while (!terminated && !_requestProcessingStopping)
+                {
+                    while (!terminated && !_requestProcessingStopping && !TakeStartLine(SocketInput))
+                    {
+                        terminated = SocketInput.RemoteIntakeFin;
+                        if (!terminated)
+                        {
+                            await SocketInput;
+                        }
+                    }
+
+                    while (!terminated && !_requestProcessingStopping && !TakeMessageHeaders(SocketInput, _requestHeaders))
+                    {
+                        terminated = SocketInput.RemoteIntakeFin;
+                        if (!terminated)
+                        {
+                            await SocketInput;
+                        }
+                    }
+
+                    if (!terminated && !_requestProcessingStopping)
+                    {
+                        MessageBody = MessageBody.For(HttpVersion, _requestHeaders, this);
+                        _keepAlive = MessageBody.RequestKeepAlive;
+                        RequestBody = new FrameRequestStream(MessageBody);
+                        ResponseBody = new FrameResponseStream(this);
+                        DuplexStream = new FrameDuplexStream(RequestBody, ResponseBody);
+
+                        var context = _application.CreateContext(this);
+                        try
+                        {
+                            await _application.ProcessRequestAsync(context).ConfigureAwait(false);
+                        }
+                        catch (Exception ex)
+                        {
+                            ReportApplicationError(ex);
+                            _application.DisposeContext(context, ex);
```

Comment:
I think this is disposing the context too soon. `FireOnCompleted` is yet to run.